eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
    * @param model to add the annotation to.
    * @param offset start of the annotation to add.
    * @param length length of the annotation.
    * @param source of the annotation.
    */
-  public void insertAnnotation(IAnnotationModel model, int offset, int length, User source) {
+  public void insertAnnotation(
+      final IAnnotationModel model, final int offset, final int length, final User source) {
 
-    if (!contribtionAnnotationsEnabled) return;
+    // Return if length == 0, (called after a deletion was performed)
+    if (!contribtionAnnotationsEnabled || length <= 0) return;
 
-    if (length < 0) // why is 0 len allowed ?
-    return;
+    if (length > 1) {
+      // Correct current model state to match with the assumption that all
+      // annotations have length &lt;= 1.
+      enforceAnnotationsWithLengthOne(model);
+    }
 
-    /* Return early if there already is an annotation at that offset */
-    for (Iterator<Annotation> it = model.getAnnotationIterator(); it.hasNext(); ) {
-      Annotation annotation = it.next();
+    final ContributionAnnotationHistory history = getHistory(source);
 
-      if (annotation instanceof ContributionAnnotation
-          && model.getPosition(annotation).includes(offset)
-          && ((ContributionAnnotation) annotation).getSource().equals(source)) {
-        return;
+    final Map<ContributionAnnotation, Position> annotationsToAdd =
+        createAnnotationsForContributionRange(model, offset, length, source);
+
+    final Pair<IAnnotationModel, List<ContributionAnnotation>> annotationsToRemove =
+        history.addNewEntry(new ArrayList<>(annotationsToAdd.keySet()));
+
+    // Case 1: An insert operation is performed and the history is full
+    if (annotationsToRemove != null) {
+      final IAnnotationModel annotationsToRemoveModel = annotationsToRemove.getLeft();
+
+      if (annotationsToRemoveModel.equals(model)) {
+        annotationModelHelper.replaceAnnotationsInModel(
+            annotationsToRemoveModel, annotationsToRemove.getRight(), annotationsToAdd);
+      } else {
+        annotationModelHelper.replaceAnnotationsInModel(
+            annotationsToRemoveModel, annotationsToRemove.getRight(), Collections.emptyMap());
+        annotationModelHelper.replaceAnnotationsInModel(
+            model, Collections.emptyList(), annotationsToAdd);
       }
     }
-
-    addContributionAnnotation(
-        new ContributionAnnotation(source, model), new Position(offset, length));
+    // Case 2: An insert operation is performed and the history is not full
+    else {
+      annotationModelHelper.replaceAnnotationsInModel(
+          model, Collections.emptyList(), annotationsToAdd);
+    }
   }
 
   /**
-   * Splits the contribution annotation at given position, so that a following text change will not
-   * expand the annotation. This needs to be called before the text is changed.
+   * This method is only called if an annotation with length &gt; 1 should be inserted, because if
```

Comment:
In actual javadoc, this style of using `<` and `>` technically works when viewed through a tool interpreting the javadoc but is still unnecessarily hard to read when looking at the code.
```suggestion
   * This method is only called if an annotation with length > 1 should be inserted, because if
```

Reply:
Hmm, I use this, because the javadoc processor complains about using "<"/">". I would like to have a javadoc build without multiple errors in the future.

Reply:
I guess then keep it the way it is for the javadoc. The normal comments should still be adjusted in my opinion. And in the javadoc, you could consider writing it out instead. But that is a matter of taste.